libs/product/src/selectors/composite-product/composite-product.selectors.ts

```diff
@@ -357,3 +128,92 @@ function getMaximumRequiredCompositeItemDiscountedPrice(item: DaffCompositeProdu
+ */
+function getMinPricesForConfiguration(product: DaffCompositeProduct, appliedOptions: Dictionary<DaffCompositeProductItemOption>): DaffProductPrices {
+	return {
+		discountedPrice: product.items.reduce((acc, item) => daffAdd(
+			acc, 
+			appliedOptions && appliedOptions[item.id] ? 
+				daffMultiply(
+					appliedOptions[item.id].discount ? daffSubtract(appliedOptions[item.id].price, appliedOptions[item.id].discount.amount) : appliedOptions[item.id].price, 
+					appliedOptions[item.id].quantity
+				) : 
+				getMinimumRequiredCompositeItemDiscountedPrice(item)
+		), product.discount ? daffSubtract(product.price, product.discount.amount) : product.price),
+		discount: { amount: null, percent: null },
+		originalPrice: product.items.reduce((acc, item) => daffAdd(
+			acc, 
+			appliedOptions && appliedOptions[item.id] ? 
+				daffMultiply(
+					appliedOptions[item.id].price, 
+					appliedOptions[item.id].quantity
+				) : getMinimumRequiredCompositeItemPrice(item)
+		), product.price)
+	}
+}
+
+/**
+ * Gets the maximum prices of a composite product for the configuration provided.
+ * @param product a DaffCompositeProduct
+ * @param appliedOptions a Dictionary<DaffCompositeProductItemOption> that determines the current configuration of the composite product.
+ */
+function getMaxPricesForConfiguration(product: DaffCompositeProduct, appliedOptions: Dictionary<DaffCompositeProductItemOption>): DaffProductPrices {
+	return {
+		discountedPrice: product.items.reduce((acc, item) => daffAdd(
+			acc, 
+			appliedOptions && appliedOptions[item.id] ? 
+				daffMultiply(
+					appliedOptions[item.id].discount ? daffSubtract(appliedOptions[item.id].price, appliedOptions[item.id].discount.amount) : appliedOptions[item.id].price, 
+					appliedOptions[item.id].quantity
+				) : 
+				getMaximumRequiredCompositeItemDiscountedPrice(item)
+		), product.discount ? daffSubtract(product.price, product.discount.amount) : product.price),
+		discount: { amount: null, percent: null },
+		originalPrice: product.items.reduce((acc, item) => daffAdd(
+			acc,
+			appliedOptions && appliedOptions[item.id] ? 
+				daffMultiply(
+					appliedOptions[item.id].price, 
+					appliedOptions[item.id].quantity
+				) : getMaximumRequiredCompositeItemPrice(item)
+		), product.price)
+	}
+}
+
+/**
+ * Gets the maximum prices of a composite product including optional item prices.
+ * @param product a DaffCompositeProduct
+ */
+function getMaxPricesIncludingOptionalItems(product: DaffCompositeProduct): DaffProductPrices {
+	return {
+		discountedPrice: (<DaffCompositeProduct>product).items.reduce((acc, item) => daffAdd(
+			acc, 
+			Math.max(...item.options.map(option => option.discount ? daffSubtract(option.price, option.discount.amount) : option.price))
+		), product.discount ? daffSubtract(product.price, product.discount.amount) : product.price),
```

Comment:
Can you extract `product.discount ? daffSubtract(product.price, product.discount.amount) : option.price` into its own function? This logic is used multiple times in this file for both options and products.

Reply:
Note that you should also use this extracted function in `getMaximumRequiredCompositeItemDiscountedPrice` and `getMinimumRequiredCompositeItemDiscountedPrice`. So for example: 

```ts
function getDiscountablePrice(product: DaffProduct): number {
  return product.discount ? daffSubtract(product.price, product.discount.amount) : product.price
}

function getMaximumRequiredCompositeItemDiscountedPrice(item: DaffCompositeProductItem): number {
  return item.required ? Math.max(...item.options.map(getDiscountablePrice)) : 0;
}
```

Reply:
done. And thanks for the extra note. Definitely would have missed that

Reply:
You still need to change `getMaximumRequiredCompositeItemDiscountedPrice` and `getMinimumRequiredCompositeItemDiscountedPrice`.

Reply:
Oh nice. Done

Reply:
no line 119 and 129.

Reply:
sorry about that. I think that's the last ones